feat: checkup CKA_ID of kek in hsm at startup

IceManGreen opened the pull request:
- checks both kek CKA_ID and CKA_LABEL during the k8s-kms-plugin at startup to avoid errors during runtime
- new arguments names consistency for kek
- update markdown documentation with new arguments consistency 
- refactoring of code for kek label and id checks up

IceManGreen committed:
feat: checkup CKA_ID of kek in hsm at startup

also fix arguments names consistency for kek

Signed-off-by: Louis Cailliot <louis.cailliot@thalesgroup.com>

Nicolas-Peiffer committed:
create a base64 output for the Encrypt response

this can then be used to test a serve rotation

IceManGreen committed:
Merge branch 'update-to-kms-api-v2' into update-to-kms-api-v2-force-cka-id

Nicolas-Peiffer commented on docs/cli-user-interface/markdown/README.md:
```diff
@@ -2,9 +2,9 @@
-- version `v0.6.0-alpha-14-g2c6222b-dirty`
-- commit `2c6222bde64053b12f739af58c1cee97c32863ec`
-- build date 2025-07-31T09:35:04+00:00.
+- version ``
+- commit ``
+- build date .
```
Version & commit info are missing, this is caused by missing `LDFLAGS` values when running `@go run cmd/k8s-kms-plugin/main.go docs ...` command in the Makefile compared to running this `docs` from the k8s-kms-plugin's binary.

One reason behind including the "version", "date" and "commit ID" in the auto-generated doc is to be able to know _for sure_ to which version of the tool this doc corresponds.

Drawback of this method is that the commit ID referenced here is obviously an older commit, since you need a new commit to commit the documentation.

We could consider a new way to populate those fields. This requires changes in the cobra command `cmd/k8s-kms-plugin/cmd/docs.go`.

Nicolas-Peiffer commented on docs/softhsm-v2.md:
```diff
@@ -117,7 +118,7 @@ k8s-kms-plugin \
     --p11-label mylabel \
     --p11-pin mypin \
-    --kek-id  aa22334455bc \
+    --p11-key-id aa22334455bc \
```
Should we modify the value and content `aa22334455bc` to match the command `--id $(head -c 8 /dev/urandom | xxd -p)` ?

Indeed, output of the 8 bytes looks like this
```
head -c 8 /dev/urandom | xxd -p
72d966d1b97e781d
```

Whereas `aa22334455bc` is 6 bytes long

IceManGreen replied:
No because here is an argument of the command `k8s-kms-plugin serve` so it has to be explicitly given, and is consistent with the documentation above

Nicolas-Peiffer replied:
What I meant was : 
* Command `head -c 8 /dev/urandom | xxd -p)` creates an 8 bytes long
* `aa22334455bc`, which is used as an example of a key id, is 6 bytes long

Maybe some readers could be bothered by the fact that on the one end `pkcs11-tool ... --id $(head -c 8 /dev/urandom | xxd -p)` will create an 8 bytes long key ID. But then in the doc we show a 6 bytes long key ID. I would make sure the key ID we use as example is 8 bytes long to match the length.

In other word:
* either use 6 bytes long Key IDs like `aa22334455bc` as an example and in the pkcs11-tool script (`head -c 6 /dev/urandom | xxd -p`)
* or use 8 bytes long Key IDs as example and in the pkcs11-tool script (`head -c 8 /dev/urandom | xxd -p`)

I is only to prevent a question from a reader "why do they use a 6 bytes `aa22334455bc` key ID when they generate a 8 bytes urandom above ?".

:arrow_forward:  **My suggestion**: replace `aa22334455bc` with `72d966d1b97e781d` :smile:  (or any 8 bytes long urandom generated string)

Nicolas-Peiffer commented on docs/cli-user-interface/markdown/README.md:
```diff
@@ -2,9 +2,9 @@
-- version `v0.6.0-alpha-14-g2c6222b-dirty`
-- commit `2c6222bde64053b12f739af58c1cee97c32863ec`
-- build date 2025-07-31T09:35:04+00:00.
+- version ``
+- commit ``
+- build date .
```
Version & commit info are missing, this is caused by missing `LDFLAGS` values when running `@go run cmd/k8s-kms-plugin/main.go docs ...` command in the Makefile compared to running this `docs` from the k8s-kms-plugin's binary.

One reason behind including the "version", "date" and "commit ID" in the auto-generated doc is to be able to know _for sure_ to which version of the tool this doc corresponds.

Drawback of this method is that the commit ID referenced here is obviously an older commit, since you need a new commit to commit the documentation.

We could consider a new way to populate those fields. This requires changes in the cobra command `cmd/k8s-kms-plugin/cmd/docs.go`.

IceManGreen replied:
Fixed at least in the makefile for the command 'doc' to include LD_FLAGS, thank you.
Concerning the commit ID issue, I tend to think that it is a matter of CI/CD execution and that it will be solved by regenerating the docs when we release a new version based on a tag commit. 
What do you think of that ?

Nicolas-Peiffer replied:
Hum :thinking: in fact the issue will probably be the same with CICD. If we tag a release with a particular tag, we will still be at least 1 commit away from commiting the updated doc that has been auto generated.

Let us address this later.

Nicolas-Peiffer commented on pkg/providers/p11.go:
```diff
@@ -791,6 +748,8 @@ func (p *P11) decryptWithContext(req *k8skmsv2.DecryptRequest, isRotation bool)
 
 			// create decryptor
 			decryptor := gose.NewJweRsaKeyEncryptionDecryptorImpl(store)
+			// TODO deleteme
```
Missing TODO to remove?

Nicolas-Peiffer commented on pkg/providers/p11.go:
```diff
@@ -1025,8 +989,18 @@ func (p *P11) Status(ctx context.Context, request *k8skmsv2.StatusRequest) (stat
 
-	// TODO: consider only testing the length of the KEK ID and not the nil check
 	if len(p.kekCkaId) == 0 {
+		var e error
+		secretKey, e := p.ctx.FindKey(nil, []byte("aes00softhsm"))
+		if e != nil {
+			print(e.Error())
+		}
+		print(secretKey)
+		a, myerr := p.ctx.GetAttribute(secretKey, pkcs11.CKA_ID)
+		if myerr != nil {
+			print(myerr.Error())
+		}
+		print(string(a.Value))
```
It looks like this snippet of code is only for testing purpose ? Should this be removed ?
Indeed, `aes00softhsm` is a hardcoded CKA_LABEL.

IceManGreen replied:
indeed, removed.

Nicolas-Peiffer commented on cmd/k8s-kms-plugin/cmd/serve.go:
```diff
@@ -232,8 +232,8 @@ func init() {
-	serveCmd.PersistentFlags().String("kek-id", "", "Key ID CKA_ID for KMS KEK. Env var: K8S_KMS_PLUGIN_SERVE_KEK_ID")
-	serveCmd.PersistentFlags().String("hmac-id", "", "Key ID CKA_ID for KMS HMAC. Env var: K8S_KMS_PLUGIN_SERVE_HMAC_ID")
+	serveCmd.PersistentFlags().String("p11-key-id", "", "Key ID CKA_ID for KMS KEK. Env var: K8S_KMS_PLUGIN_SERVE_KEK_ID")
+	serveCmd.PersistentFlags().String("p11-hmac-id", "", "Key ID CKA_ID for KMS HMAC. Env var: K8S_KMS_PLUGIN_SERVE_HMAC_ID")
```
Need to fix the env vars in the Cobra command description field.

* `K8S_KMS_PLUGIN_SERVE_KEK_ID` becomes `K8S_KMS_PLUGIN_SERVE_P11_KEY_ID`
* `K8S_KMS_PLUGIN_SERVE_HMAC_ID` becomes `K8S_KMS_PLUGIN_SERVE_P11_HMAC_ID`

Or consider removing env vars from the cobra description field.

IceManGreen replied:
I suggest to remove it from the description since it already appears in the 'env var' column of the generated tables.
Keep in them in the "usage" column, manually, induces more maintenance and is prone to potential human mistakes for future releases.

Nicolas-Peiffer replied:
I agree, it is too much manual maintenance when we change a tag name. Lets remove Env Var from the cobra "usage/description".

Nicolas-Peiffer commented on cmd/k8s-kms-plugin/cmd/rotation.go:
```diff
@@ -182,17 +182,17 @@ func init() {
+	rotationCmd.Flags().String("old-p11-hmac-id", "", "Key ID CKA_ID for old KEK HMAC")
 	rotationCmd.Flags().String("old-p11-hmac-label", "", "Key Label CKA_LABEL for old KEK HMAC")
-	rotationCmd.Flags().String("old-kek-id", "", "Key ID CKA_ID for old KEK")
+	rotationCmd.Flags().String("old-p11-key-id", "", "Key ID CKA_ID for old KEK")
```
Should we add env vars in the cobra command description ?

IceManGreen replied:
I tend to think no, due to my previous comment.

Nicolas-Peiffer replied:
agreed, too much maintenance. And the `docs` command gives a solution with the table with all the fields and flags

Nicolas-Peiffer commented on docs/cli-user-interface/markdown/cli-env-var-table.md:
```diff
@@ -23,11 +23,11 @@
 | k8s-kms-plugin serve | --host |  | K8S_KMS_PLUGIN_SERVE_HOST | k8s-kms-plugin.serve.host | 0.0.0.0 | string | true | Hostname without port. Env var: K8S_KMS_PLUGIN_SERVE_HOST. |
-| k8s-kms-plugin serve | --kek-id |  | K8S_KMS_PLUGIN_SERVE_KEK_ID | k8s-kms-plugin.serve.kek-id |  | string | true | Key ID CKA_ID for KMS KEK. Env var: K8S_KMS_PLUGIN_SERVE_KEK_ID |
 | k8s-kms-plugin serve | --native-path | -p | K8S_KMS_PLUGIN_SERVE_NATIVE_PATH | k8s-kms-plugin.serve.native-path | .keys | string | true | Path to key store for native provider(Files only). Env var: K8S_KMS_PLUGIN_SERVE_NATIVE_PATH. |
+| k8s-kms-plugin serve | --p11-hmac-id |  | K8S_KMS_PLUGIN_SERVE_P11_HMAC_ID | k8s-kms-plugin.serve.p11-hmac-id |  | string | true | Key ID CKA_ID for KMS HMAC. Env var: K8S_KMS_PLUGIN_SERVE_HMAC_ID |
```
Need to fix the env vars in the Cobra command description field in `serve.go` and `rotation.go`.

* `K8S_KMS_PLUGIN_SERVE_HMAC_ID` becomes `K8S_KMS_PLUGIN_SERVE_P11_HMAC_ID`

Or consider removing env vars from the cobra description field.

Nicolas-Peiffer commented on docs/cli-user-interface/markdown/cli-env-var-table.md:
```diff
@@ -23,11 +23,11 @@
 | k8s-kms-plugin serve | --native-path | -p | K8S_KMS_PLUGIN_SERVE_NATIVE_PATH | k8s-kms-plugin.serve.native-path | .keys | string | true | Path to key store for native provider(Files only). Env var: K8S_KMS_PLUGIN_SERVE_NATIVE_PATH. |
+| k8s-kms-plugin serve | --p11-hmac-id |  | K8S_KMS_PLUGIN_SERVE_P11_HMAC_ID | k8s-kms-plugin.serve.p11-hmac-id |  | string | true | Key ID CKA_ID for KMS HMAC. Env var: K8S_KMS_PLUGIN_SERVE_HMAC_ID |
 | k8s-kms-plugin serve | --p11-hmac-label |  | K8S_KMS_PLUGIN_SERVE_P11_HMAC_LABEL | k8s-kms-plugin.serve.p11-hmac-label |  | string | true | Key Label CKA_LABEL to use for sha based verifications. Env var: K8S_KMS_PLUGIN_SERVE_P11_HMAC_LABEL. |
+| k8s-kms-plugin serve | --p11-key-id |  | K8S_KMS_PLUGIN_SERVE_P11_KEY_ID | k8s-kms-plugin.serve.p11-key-id |  | string | true | Key ID CKA_ID for KMS KEK. Env var: K8S_KMS_PLUGIN_SERVE_KEK_ID |
```
Need to fix the env vars in the Cobra command description field in `serve.go`.

* `K8S_KMS_PLUGIN_SERVE_KEK_ID` becomes `K8S_KMS_PLUGIN_SERVE_P11_KEY_ID`

Or consider removing env vars from the cobra description field.

Nicolas-Peiffer reviewed (changes requested):
I build and tested against the grpcurl scripts for both `serve` and `serve rotation`.

I notice only a couple documentation fixes to make before merging.

Nicolas-Peiffer committed:
Merge branch 'update-to-kms-api-v2' into update-to-kms-api-v2-force-cka-id

Nicolas-Peiffer commented:
@IceManGreen  We also need to update the YAML [configs/config.example.yaml](https://github.com/ThalesGroup/k8s-kms-plugin/blob/55868c5b531b07d00ef63eb9f73edfa78fef4705/configs/config.example.yaml#L34) to update the YAML keys/values with new names.

We need to verify the content of the sample `configs/config.example.yaml` by comparing to `docs/cli-user-interface/txt/cli-env-var-table.txt` or to `docs/cli-user-interface/markdown/cli-env-var-table.md`.

For example, we need to modify at least those YAML keys:

```yaml
k8s-kms-plugin:
  # Cobra subcommands
  serve:
    kek-id: "my-key"
    p11-label: "my-current-label"
```

Becomes

```yaml
k8s-kms-plugin:
  # Cobra subcommands
  serve:
    p11-key-id: "72d966d1b97e781d"
    p11-key-label: "my-current-label"
    p11-hmac-id: "..."
    p11-hmac-label: "..."
```

And I notice other YAML keys that needs to be adapted to latest cobra flags.

IceManGreen committed:
fix: address pr #74 comments

Signed-off-by: Louis Cailliot <louis.cailliot@thalesgroup.com>

IceManGreen committed:
Merge branch 'update-to-kms-api-v2' into update-to-kms-api-v2-force-cka-id

IceManGreen commented:
@Nicolas-Peiffer I addressed most of your comments.
I left some discussion pending, waiting for your opinion.
I changed the config yaml example just a bit, to comment the optional arguments and left only the most used and mandatory ones. Tell me what you think about it :)
